src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/custom_vehicle_data_manager_impl.cc

```diff
@@ -0,0 +1,535 @@
+      LOG4CXX_WARN(logger_,
+                   "Float value: " << value
+                                   << " does not fit into boundaries : ("
+                                   << minval << " , " << maxval << ")");
+    }
+    return result;
+  }
+  if (item_type == policy_table::VehicleDataItem::kBoolean) {
+    if (smart_objects::SmartType::SmartType_Boolean != item.getType()) {
+      LOG4CXX_WARN(logger_,
+                   "Parameter type mismatch. Expected: "
+                       << policy_table::VehicleDataItem::kBoolean
+                       << ", actual: " << item.getType());
+      return false;
+    }
+  }
+  return result;
+}
+
+bool CustomVehicleDataManagerImpl::ValidateStructTypeItem(
+    const ns_smart_device_link::ns_smart_objects::SmartObject& item,
+    const policy_table::VehicleDataItem& item_schema) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  auto find_param = [&item](const std::string& name) {
+    for (auto param : item.enumerate()) {
+      if (param == name) {
+        return item[param];
+      }
+    }
+    return smart_objects::SmartObject();
+  };
+
+  // Check for redundant parameters
+  for (const auto& param : item.enumerate()) {
+    auto param_schema = FindSchemaByNameRecursive(param);
+    if (!param_schema.is_initialized()) {
+      return false;
+    }
+  }
+
+  auto params_schemas = *(item_schema.params);
+  for (const auto& param_schema : params_schemas) {
+    smart_objects::SmartObject param =
+        find_param(std::string(param_schema.name));
+    if (smart_objects::SmartType::SmartType_Null == param.getType()) {
+      if (param_schema.mandatory) {
+        return false;
+      }
+      continue;
+    }
+    if (!ValidateVehicleDataItem(param, param_schema)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename Comparer>
+const OptionalDataItem FindSchema(
+    const std::vector<policy_table::VehicleDataItem>& oem_items,
+    SearchMethod search_method,
+    Comparer comparer) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  std::vector<policy_table::VehicleDataItem> items;
+  for (const auto& item : oem_items) {
+    if (comparer(item)) {
+      items.push_back(item);
+    }
+    if (search_method == SearchMethod::RECURSIVE &&
+        policy_table::VehicleDataItem::kStruct == std::string(item.type)) {
+      auto found_in_params =
+          FindSchema(*(item.params), search_method, comparer);
+      if (found_in_params.is_initialized()) {
+        items.push_back(*found_in_params);
+      }
+    }
+  }
+  std::sort(items.begin(),
+            items.end(),
+            [](const policy_table::VehicleDataItem& left,
+               const policy_table::VehicleDataItem& right) {
+              if (!right.since.is_initialized()) {
+                return false;
+              }
+              if (!left.since.is_initialized()) {
+                return true;
+              }
+              const std::string l = *left.since;
+              const std::string r = *right.since;
```

Comment:
@jacobkeeler currently RPCspec versioning does not support negotiating parameters validation according to application message version. 
Validation of custom parameters done according to RPC spec current state.

Reply:
What do you mean? Currently any normal mobile requests are validated based on the negotiated message version, this is why a version is passed into the [validate function](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/smart_objects/include/smart_objects/smart_schema.h#L78). Admittedly this might not be possible to do with notifications since they are often constructed to broadcast to several apps.

Reply:
Will be is not actual as soon as dynamic parameters will be added to validation schema.
Implementation will be ready for Friday 23 May

Reply:
Please see updates in commit 947c25585786f442a296fd4375c50d892684336b